src/oidcendpoint/oidc/token.py

```diff
@@ -83,26 +118,91 @@ def _access_token(self, req, **kwargs):
         issue_refresh = False
         if "issue_refresh" in kwargs:
             issue_refresh = kwargs["issue_refresh"]
+        else:
+            if "offline_access" in grant.scope:
+                issue_refresh = True
+
+        _response = {
+            "token_type": "Bearer",
+            "expires_in": 900,
+            "scope": grant.scope,
+            "state": _authn_req["state"]
+        }
+
+        token = self._mint_token("access_token",
+                                 grant,
+                                 _session_info["session_id"],
+                                 _session_info["client_id"],
+                                 grant.sub,
+                                 code)
+        _response["access_token"] = token.value
+
+        if issue_refresh:
+            refresh_token = self._mint_token("refresh_token", grant,
+                                             _session_info["session_id"],
+                                             _session_info["client_id"],
+                                             grant.sub,
+                                             code)
+            _response["refresh_token"] = refresh_token.value
+
+        code.register_usage()
```

Comment:
This, unfortunately, does not work if the session db is not an in-memory dict. Every attribute change should use the get/set database interface.

Reply:
Fix been pushed.

Reply:
Thanks!

Reply:
Unfortunately, there's still an issue here when using a database. The grant that is updated does not contain the code object that just got changed (with `register_usage`). The reason is the methods in line 92 and 95 (`get_session_info_by_token` and `find_token`).

More specifically, `get_session_info_by_token` fetches the whole session_info (including the issued_tokens). After that instead of looking into the fetched objects and finding the right code, we use `find_token` which AGAIN looks into the database and finds the correct token. As a result, the same code is fetched but now it is a different object because the `to_json` method of course creates a new one. 

I recommend we change the `find_token` method or create a new one that instead of querying the database, is given a session_info and tries to find a token inside the object. That way we both solve our bug and reduce the queries needed in the database.

Reply:
The `grant` in `find_token` could be an optional param, when that is `None` we query the database:
```python
    def find_token(self, token_value: str, session_id: str = None, grant: Grant = None) -> Optional[Token]:
        """

        :param session_id: Based on 3-tuple, user_id, client_id and grant_id
        :param token_value:
        :param grant: A grant object
        :return:
        """
        if not grant:
                if not session_id:
                        raise ValueError("One of `grant` or `session_id` must be provided")
                user_id, client_id, grant_id = unpack_session_key(session_id)
                grant = self.get([user_id, client_id, grant_id])
        for token in grant.issued_token:
            if token.value == token_value:
                return token

        return None
```

Reply:
Unfortuanately, this still does not work for databases.

In line 92 `get_session_info_by_token`  is called so the session_info is fetched from the DB and created with `from_json`. After that, `find_token` is called in line 95 and the database is again queried (although we already have the token inside session_info's issued_tokens). As a result, the code returned is again created with `from_json` and although it's the same as the one inside session_info, it's a different object. 

So when `code.register_usage()` is called the code object changes (self.used is incremented) but the code inside `session_info` stays unchanged.

This could be solved if instead of using `find_token` (and querying the database a second time without reason) we could fetch the code from inside `session_info` 

         _session_info = _mngr.get_session_info_by_token(_access_code, grant=True)
        code = _mngr.find_token(_session_info["session_id"], _access_code)